sched: Remove multi-level queue priority limit

TaiJuWu opened the pull request:
Modified bitmask to  bitmask array, it can make multilevel queue remove
32 bit prioriry limit.

We can scan bitmask array to find which queue have ready thread.

Only need the number of queues as priority because the priority
is checked on create_thread.

TaiJuWu commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
 #include <zephyr/sys/dlist.h>
 #include <zephyr/sys/rb.h>
 
+#define K_HIGHEST_THREAD_PRIO             (-CONFIG_NUM_COOP_PRIORITIES)
```
This part already defined in <zephyr/kernel.h> but this file can't be included in this file, it will cause compiler error.
Does anyone have suggestions about this code?

andyross reviewed (changes requested):
I don't normally make a big deal over style, but there are a LOT of whitespace-only changes here.  Can you split those out so that the logic changes are isolated?  I'm fine if you want to rework the whitespace in a separate patch.

No complaints in general about the change, but it's difficult to review in its current state

TaiJuWu commented:
> I don't normally make a big deal over style, but there are a LOT of whitespace-only changes here. Can you split those out so that the logic changes are isolated? I'm fine if you want to rework the whitespace in a separate patch.
> 
> No complaints in general about the change, but it's difficult to review in its current state



> I don't normally make a big deal over style, but there are a LOT of whitespace-only changes here. Can you split those out so that the logic changes are isolated? I'm fine if you want to rework the whitespace in a separate patch.
> 
> No complaints in general about the change, but it's difficult to review in its current state

I apology for this and already split logic and the result of format tool.

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
+#define K_IDLE_PRIO K_LOWEST_THREAD_PRIO
+#define K_HIGHEST_APPLICATION_THREAD_PRIO (K_HIGHEST_THREAD_PRIO)
+#define K_LOWEST_APPLICATION_THREAD_PRIO (K_LOWEST_THREAD_PRIO - 1)
+#define K_THRAD_PRIO_NUM (K_LOWEST_THREAD_PRIO - K_HIGHEST_THREAD_PRIO + 1)
```
Spelling.  Also not really a fan of adding this as an API, maybe move this to the C file and call it "NUM_PRIO" or something simpler?

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
 #include <zephyr/sys/dlist.h>
 #include <zephyr/sys/rb.h>
 
+#define K_HIGHEST_THREAD_PRIO (-CONFIG_NUM_COOP_PRIORITIES)
```
Indeed, this is duplicated from kernel.h.  If you can't make the header cycles work out from that spot, the solution would be to move it to a scheduler header with no/minimal dependencies and include it in both places.

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
+#define PRI_BITMAP_ITERATOR PRI_QUEUE_BITMAP_SIZE
+#define BIT_NUM 32
+#define BIT_ARRAY_IDX(prio) ((prio) / BIT_NUM)
+#define BIT_ARRAY_OFFSET(prio) ((prio) % BIT_NUM)
```
Likewise these probably want to be in sched.c and not in a header included via the public API

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -51,8 +63,8 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq);
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[PRI_QUEUE_BITMAP_SIZE * BIT_NUM];
+	unsigned int bitmask[PRI_QUEUE_BITMAP_SIZE]; /* bit 1<<i set if queues[i] is non-empty */
```
You don't need the full bitmap size, just the number of actual priorities.  Also the comment is now confusing as it doesn't specify which array element comes first.

andyross commented on kernel/sched.c:
```diff
@@ -1338,44 +1338,49 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
-	int priority_bit = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int offset_priority = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	int priority_idx = BIT_ARRAY_IDX(offset_priority);
+	int priority_bit = BIT_ARRAY_OFFSET(offset_priority);
```
Style: you repeat this construction in a few places.  Might be cleaner to define a utility that returns both a word pointer and a bit index for a given priority.

cfriedt replied:
Yep - a static (possibly inline) function in this file would do it.

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
+#define K_LOWEST_APPLICATION_THREAD_PRIO (K_LOWEST_THREAD_PRIO - 1)
+#define K_THRAD_PRIO_NUM (K_LOWEST_THREAD_PRIO - K_HIGHEST_THREAD_PRIO + 1)
+#define PRI_QUEUE_BITMAP_SIZE ((K_THRAD_PRIO_NUM + BIT_NUM - 1) / BIT_NUM)
+#define PRI_BITMAP_ITERATOR PRI_QUEUE_BITMAP_SIZE
```
Confusing name, and needless.  "BITMAP_SIZE" makes more sense everywhere this is used.

andyross commented on kernel/sched.c:
```diff
@@ -1338,44 +1338,49 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
+		sys_dnode_t *n = sys_dlist_peek_head(l);
+
+		if (n != NULL) {
+			thread = CONTAINER_OF(n, struct k_thread, base.qnode_dlist);
```
This logic looks wrong?  If you find a thread in the first word, you'll continue searching in the other words (which hold lower priorities!) and return the wrong thread.  You want to return right here, no?

TaiJuWu replied:
You're right. It's wrong. I will fixed it later.

andyross commented on kernel/sched.c:
```diff
@@ -1402,6 +1407,9 @@ void init_ready_q(struct _ready_q *rq)
 #elif defined(CONFIG_SCHED_MULTIQ)
+	for (int i = 0; i < PRI_QUEUE_BITMAP_SIZE; ++i) {
+		_kernel.ready_q.runq.bitmask[i] = 0;
+	}
```
The kernel struct is zero-filled at startup, no need for this.

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -51,8 +63,8 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq);
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[PRI_QUEUE_BITMAP_SIZE * BIT_NUM];
+	unsigned int bitmask[PRI_QUEUE_BITMAP_SIZE]; /* bit 1<<i set if queues[i] is non-empty */
```
Also, while you're in here: it would be good to:

* Make this a stdint type, which is the current Zephyr style, and not an "unsigned int"
* Allow for 64 bit platforms to use uint64_t here (and __builtin_clzll() of course) for efficiency.

TaiJuWu replied:
Is there a method that I can know my machine is 32bit or 64 bit?

cfriedt replied:
`uintptr_t` would do it for both.

cfriedt replied:
or `long`

TaiJuWu replied:
Done.

andyross commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,18 @@
+#define K_THRAD_PRIO_NUM (K_LOWEST_THREAD_PRIO - K_HIGHEST_THREAD_PRIO + 1)
+#define PRI_QUEUE_BITMAP_SIZE ((K_THRAD_PRIO_NUM + BIT_NUM - 1) / BIT_NUM)
+#define PRI_BITMAP_ITERATOR PRI_QUEUE_BITMAP_SIZE
+#define BIT_NUM 32
```
Generally better to compute this via sizeof() at the spot where it's used, e.g. `8 * sizeof(pq->bitmask[0]` or whatever.

TaiJuWu replied:
`#define BIT_NUM (8 * sizeof(((struct _priq_mq *)0)->bitmask[0]))` will cause `invalid use of undefined type 'struct _priq_mq'`

andyross reviewed (commented):
More notes.  Definitely check the logic error in the priority search (or explain to me how I'm misinterpreting).

Also: this will need to update the documentation for MULTIQ in both the KConfig file and the .rst docs to remove the text talking about the 32 priority limitation.

cfriedt commented on tests/kernel/sched/schedule_api/prj_multiq.conf:
```diff
@@ -5,3 +5,5 @@ CONFIG_SCHED_MULTIQ=y
 CONFIG_MAX_THREAD_BYTES=5
 CONFIG_MP_MAX_NUM_CPUS=1
 CONFIG_ZTEST_FATAL_HOOK=y
+CONFIG_NUM_COOP_PRIORITIES=20
+CONFIG_NUM_PREEMPT_PRIORITIES=20
```
Is this mainly just to check that this will work when there are > 32 total priorities?

It might not be necessary to explicitly do that, since the `BIT_*()` macros should already be tested to work in those scenarios (i.e. there may be coverage already)

TaiJuWu replied:
> Is this mainly just to check that this will work when there are > 32 total priorities?
> 
Yes, it is.
> It might not be necessary to explicitly do that, since the `BIT_*()` macros should already be tested to work in those scenarios (i.e. there may be coverage already)

Ok. Thanks for your information.

TaiJuWu replied:
Hi @cfriedt,
I have a question which is how can I check this feature be tested?

cfriedt commented on doc/kernel/services/scheduling/index.rst:
```diff
@@ -82,7 +82,11 @@ runtime overhead and performance scaling when many threads are added.
 
   When selected, the scheduler ready queue will be implemented as the
-  classic/textbook array of lists, one per priority (max 32 priorities).
+  classic/textbook array of lists, one per priority.
+
+  This scheduler will invoke mod operation so it is suggested
+  to set (:kconfig:option:`CONFIG_NUM_PREEMPT_PRIORITIES`) + (:kconfig:option:`CONFIG_NUM_COOP_PRIORITIES`)
+  to power of 2 minus 1 (e.g. 31, 63, 127...).
```
It should be fine to set either `CONFIG_NUM_PREEMPT_PRIORITIES` or `CONFIG_NUM_COOP_PRIORITIES` to a value that is not a power of two minus one as well.

From what I understand, there can be a maximum of 256 priorities in total (including both preempt and coop) based on this

https://github.com/zephyrproject-rtos/zephyr/blob/840db7858e90cb49c1ba253ce037a44c01549b18/include/zephyr/kernel/thread.h#L95

I'm not sure if we need more than 256, to be honest.

cfriedt replied:
It should be fine to set either `CONFIG_NUM_PREEMPT_PRIORITIES` or `CONFIG_NUM_COOP_PRIORITIES` to a value that is not a power of two minus one as well.

From what I understand, there can be a maximum of 256 priorities in total (including both preempt and coop) based on this

https://github.com/zephyrproject-rtos/zephyr/blob/840db7858e90cb49c1ba253ce037a44c01549b18/include/zephyr/kernel/thread.h#L95

I'm not sure if we need more than 256, to be honest.

Maybe this could be more short and sweet?
```suggestion
  classic/textbook array of lists, one per priority.
```

cfriedt commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -10,6 +10,10 @@
 #include <zephyr/sys/dlist.h>
 #include <zephyr/sys/rb.h>
 
+#define NBITS (8 * sizeof(uintptr_t))
+#define K_NUM_THRAD_PRIO (CONFIG_NUM_PREEMPT_PRIORITIES + CONFIG_NUM_COOP_PRIORITIES + 1)
+#define PRI_QUEUE_BITMAP_SIZE ((K_NUM_THRAD_PRIO + NBITS - 1) / NBITS)
+
```
This (or something like it) should probably go into sched.c file, as Andy mentioned before.

TaiJuWu replied:
It can’t. Because the  _prio_q  will use this macro.
I already minimize the macros.

cfriedt commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -44,15 +48,14 @@ void z_priq_rb_remove(struct _priq_rb *pq, struct k_thread *thread);
 struct _priq_mq {
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[PRI_QUEUE_BITMAP_SIZE * NBITS];
+	uintptr_t bitmask[PRI_QUEUE_BITMAP_SIZE];
```
What do you think of the following @TaiJuWu ?

```cpp
#define NUM_PRIO (CONFIG_NUM_PREEMPT_PRIORITIES + CONFIG_NUM_COOP_PRIORITIES)
#define PRIQ_BITMAP_SIZE (ROUND_UP(NUM_PRIO, sizeof(uintptr_t)) / sizeof(uintptr_t))
```

```suggestion
	sys_dlist_t queues[NUM_PRIO];
	uintptr_t bitmask[PRI_QUEUE_BITMAP_SIZE];
```

TaiJuWu replied:
Great. We can use NHPOT to enforce 2^x if needed.

cfriedt commented on kernel/Kconfig:
```diff
@@ -341,11 +341,13 @@ config SCHED_MULTIQ
-	  only a tiny code size overhead vs. the "dumb" scheduler and
-	  runs in O(1) time in almost all circumstances with very low
+	  as the classic/textbook array of lists, one per priority.
+	  This scheduler will invoke mod operation so it is suggested
+	  to set CONFIG_NUM_PREEMPT_PRIORITIES + CONFIG_NUM_COOP_PRIORITIES
+	  to power of 2 minus 1 (e.g. 31, 63, 127...). This corresponds
```
```suggestion
	  This corresponds
```

cfriedt commented on kernel/sched.c:
```diff
@@ -1338,45 +1338,64 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
+	struct prio_info ret;
+
+	ret.offset_prio = old_prio - K_HIGHEST_THREAD_PRIO;
+	ret.idx = ret.offset_prio / NBITS;
+	ret.bit = ret.offset_prio % NBITS;
```
Mod is an expensive operation on some architectures. Can the same thing be achieved a different way?

Do we already have something in e.g. `sys/util.h` do get the index and bit of a multi-word bitmap entry?

TaiJuWu replied:
> Mod is an expensive operation on some architectures. Can the same thing be achieved a different way?
> 
yes, there is a tradeoff.
We can enforce bitmap array is 2^x and the mod operation will be replace with &, but it will need additional memory.
This is the reason my suggestion in doc as below.
‘’’
This scheduler will invoke mod operation so it is suggested
	  to set CONFIG_NUM_PREEMPT_PRIORITIES + CONFIG_NUM_COOP_PRIORITIES
	  to power of 2 minus 1 (e.g. 31, 63, 127...).
‘’’

> Do we already have something in e.g. `sys/util.h` do get the index and bit of a multi-word bitmap entry?

I didn’t see that.

peter-mitsis commented on kernel/sched.c:
```diff
@@ -1338,45 +1338,64 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
-	struct k_thread *thread = NULL;
-	sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask)];
-	sys_dnode_t *n = sys_dlist_peek_head(l);
+	for (int i = 0; i < NUM_PRIO; ++i) {
+		if (!pq->bitmask[i]) {
+			continue;
+		}
```
`i < NUM_PRIO` is not the right termination condition (as it would in theory allow for accessing beyond the bounds of pq->bitmask[]--in practice though, I doubt that would happen). Better, I think, to use `i < PRIQ_BITMAP_SIZE`.

TaiJuWu replied:
Fixed.

peter-mitsis commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -44,15 +48,15 @@ void z_priq_rb_remove(struct _priq_rb *pq, struct k_thread *thread);
 struct _priq_mq {
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[PRIQ_BITMAP_SIZE * NBITS];
```
Question -- could we not use `sys_dlist_t queues[NUM_PRIO];` ? (Both z_setup_new_thread() and k_thread_priority_set() have a Z_ASSERT_VALID_PRIO check to verify that the priority is in range so we should not be accessing beyond the bounds of the array.)

TaiJuWu replied:
Yes, if the condition already is checked.

peter-mitsis commented on kernel/sched.c:
```diff
@@ -1338,45 +1338,64 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
+	struct prio_info ret;
+
+	ret.offset_prio = old_prio - K_HIGHEST_THREAD_PRIO;
+	ret.idx = ret.offset_prio & ~(NBITS - 1);
```
I think the calculation (and subsequent use) of the `idx` field is incorrect. As it currently stands, it is only clearing the lowest (NBITS - 1) bits. To use it later as an index into `pq->bitmask[]`, a right shift (or division) is required; otherwise, as it currently stands, when it is used as an index, it may exceed the bounds of the `bitmask[]` array.

TaiJuWu replied:
Fixed. Thank for review.

peter-mitsis commented on include/zephyr/kernel/internal/sched_priq.h:
```diff
@@ -44,15 +48,15 @@ void z_priq_rb_remove(struct _priq_rb *pq, struct k_thread *thread);
-	sys_dlist_t queues[32];
-	unsigned int bitmask; /* bit 1<<i set if queues[i] is non-empty */
+	sys_dlist_t queues[NUM_PRIO];
+	uintptr_t bitmask[PRIQ_BITMAP_SIZE];
```
I don't think that `uintptr_t` is the correct type to use here as there does not seem to be anything about the use of `bitmask[]` that suggests that pointers are necessary. My recommendation would be to use a `uint32_t` to make it clear how many bits each element in `bitmask[]` is supposed to have as `uintptr_t` varies in size.

TaiJuWu replied:
I agree with you. Using uintptr_t is for effective. In 64-bit machine we can use 64-bit `clz` operation.
You can followed [here.](https://github.com/zephyrproject-rtos/zephyr/pull/68318#discussion_r1473132254)

cfriedt replied:
Likely the optimal solution is whatever the architecture word size is, which would explain the Linux kernel 's BITS_TO_LONGS, etc.

https://elixir.bootlin.com/linux/v4.5/ident/BITS_TO_LONGS

This is one area where we could probably improve. We do have sys/bitarray.h but the issue there is that it pulls in a spinlock.

TaiJuWu replied:
In general, I prefer to use `uint32` as bitarray length even It will lost little performance; and the data type `uint32_t` is enough, more general and sys_bitarray do the same thing.
The reason that I didn't want to use sys_bitarray will pull in spinlock and it's unnecessary.

cfriedt replied:
@TaiJuWu

uint32_t is probably safe here. Could you please create an issue for bitarray and reference it (`#1234`) here in a comment?

I'm thinking
* remove spinlock (user should synchronize)
* maybe provide additional macro accessors?

Likely these are things that should probably be fixed.

TaiJuWu replied:
Sure. I will do that recently.

TaiJuWu replied:
> @TaiJuWu
> 
> uint32_t is probably safe here. Could you please create an issue for bitarray and reference it (`#1234`) here in a comment?
> 
> I'm thinking
> 
> * remove spinlock (user should synchronize)
> * maybe provide additional macro accessors?
> 
> Likely these are things that should probably be fixed and the same issues have been mentioned elsewhere.

Sure. I can do that recently.

TaiJuWu committed:
sched: Remove multi-level queue priority limit

Modified bitmask to  bitmask array, it can make multilevel queue remove
32 bit prioriry limit.

We can scan bitmask array to find which queue have ready thread.

Only need the number of queues as priority because the priority
is checked on create_thread.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

peter-mitsis commented on include/zephyr/kernel_structs.h:
```diff
@@ -32,6 +32,11 @@
 #endif
 
+#define NBITS (8 * sizeof(uint32_t))
+#define K_NUM_THRAD_PRIO (CONFIG_NUM_PREEMPT_PRIORITIES + CONFIG_NUM_COOP_PRIORITIES + 1)
```
Typo: K_NUM_THRAD_PRIO -> K_NUM_THREAD_PRIO

peter-mitsis commented on kernel/priority_queues.c:
```diff
@@ -94,16 +94,18 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
-	if (!pq->bitmask) {
-		return NULL;
-	}
+	for (int i = 0; i < K_NUM_THRAD_PRIO; ++i) {
+		if (!pq->bitmask[i]) {
```
We are iterating through the loop the wrong number of times--this has the potential to exceed the bounds of bitmask[]. Better to stop at PRIQ_BITMAP_SIZE.

andyross replied:
Yeah, this is wrong.  To state it differently: the size of the bitmask[] array is 1/32 the number of priorities, not the number of priorities.  This is going to run right off the end of the array.

peter-mitsis commented on kernel/priority_queues.c:
```diff
@@ -94,16 +94,18 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
-	struct k_thread *thread = NULL;
-	sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask)];
-	sys_dnode_t *n = sys_dlist_peek_head(l);
+		sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask[i])];
```
I don't think that we are indexing the correct queue. There does not seem to be any extra addition to account for a skipped bitmask[i]. That is, if bitmask[0] is 0, then bitmask[1] would (incorrectly) index to the same queues that bitmask[0] would.

andyross commented on kernel/include/priority_q.h:
```diff
@@ -60,27 +60,41 @@ bool z_priq_rb_lessthan(struct rbnode *a, struct rbnode *b);
+
+	ret.offset_prio = old_prio - K_HIGHEST_THREAD_PRIO;
+	ret.idx = ret.offset_prio / NBITS;
+	ret.bit = ret.offset_prio % NBITS;
```
This is clever, nice utility

andyross commented on include/zephyr/kernel_structs.h:
```diff
@@ -32,6 +32,11 @@
 #include <zephyr/sys/rb.h>
 #endif
 
+#define NBITS (8 * sizeof(uint32_t))
```
Move NBITS to the C file that uses it, or if there's more than one to a kernel-internal header somewhere.  This is a global header and tiny symbols with ambiguous names pollute the global namespace.

Also repeating the request that you make the type of the bitmask[] array be 64 bits on 64 bit platforms for efficiency.

andyross reviewed (commented):
Getting closer.  A few mistakes @peter-mitsis points out, a few new nitpicks from me.

andyross commented on kernel/priority_queues.c:
```diff
@@ -94,16 +95,22 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
-	if (n != NULL) {
-		thread = CONTAINER_OF(n, struct k_thread, base.qnode_dlist);
+		if (n != NULL) {
+			return CONTAINER_OF(n, struct k_thread, base.qnode_dlist);
```
Style: early returns have their place, but in this case it's needless as you're going to fall through to the  return anyway.  I liked the original version better, basically.

andyross reviewed (commented):
Looking very close.  I realize now it doesn't have a test case for the "second multiq word".  Can you add a case somewhere under tests/kernel/sched to exercise a configuration with 65 or more priorities?

TaiJuWu committed:
test: Remove multi level queue limit test

Set total priorities to 71.
This setting already exceeds original limit(32) and exceeds 64 to test
secondary bitmap.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

andyross reviewed (approved):
I'm out of complaints.  Looks good.  Thanks.

TaiJuWu commented:
> I'm out of complaints. Looks good. Thanks.

Thanks for review.
